### PR TITLE
[core] Prepare for material v6

### PR DESCRIPTION
**Pull request by LukasTy**
Part of #14055

Apply changes based on https://next.mui.com/material-ui/migration/migrating-to-v6/#add-styles-for-specific-color-modes.
Run the following command on `packages` and `docs` folders:
- npx @mui/codemod@next v6.0.0/styled
- npx @mui/codemod@next v6.0.0/sx-prop
- npx @mui/codemod@next v6.0.0/theme-v6
After running codemods, irrelevant changes have been removed and slight optimizations applied (i.e. replace `border` redeclaration with only overriding the `borderColor`)

I've also checked other usages of `theme.palette.mode` and tried adjusting them all (mostly docs) to play nicely with new API thus avoiding the possible dark mode flicker.

Actual changes on the code are limited, it's mostly docs.

### CODE REVIEW

**Comment by cherniavskii** [comment]
Great timing, I've just opened https://github.com/mui/mui-x/pull/14142 to run the tests against `@mui/material@^6` 😅

**Comment by LukasTy** [comment]
Yeah, I saw you opening that PR and decided to open the draft of moving to `applyStyles`. 😃

File: docs/data/date-pickers/custom-opening-button/AddWarningIconWhenInvalidRange.tsx [hunk 1]
```diff
@@ -12,7 +12,18 @@ function CustomInputAdornment(props: InputAdornmentProps & { hasError?: boolean
...
+        sx={[
+          hasError
+            ? {
+                opacity: 1,
+              }
+            : {
+                opacity: 0,
+              },
+        ]}
+      />
```
Comment (cherniavskii) [inline docs/data/date-pickers/custom-opening-button/AddWarningIconWhenInvalidRange.tsx:26]: Is this change required? It does not seem to make any difference, except that it now creates 2 objects instead of one 🤔

Comment (LukasTy) [reply docs/data/date-pickers/custom-opening-button/AddWarningIconWhenInvalidRange.tsx:26]: If I remember correctly, @siriwatknp confirmed that is is required for Pigment CSS to work, the previous syntax would not be supported by static extraction. 🤔

File: docs/src/modules/components/DemoPropsForm.tsx [hunk 2]
```diff
@@ -129,15 +129,15 @@ function ControlledColorRadio(props: any) {
           <Check
             fontSize="medium"
             color="inherit"
-            sx={{
+            sx={(theme) => ({
```
Comment (cherniavskii) [inline docs/src/modules/components/DemoPropsForm.tsx:132]: Is this for future Pigment CSS support?
We seem to miss this in the migration guide https://next.mui.com/material-ui/migration/migrating-to-pigment-css/

File: packages/x-data-grid-premium/src/components/GridGroupingColumnLeafCell.tsx [hunk 3]
```diff
@@ -9,13 +9,16 @@ function GridGroupingColumnLeafCell(props: GridRenderCellParams) {
...
-          rootProps.rowGroupingColumnMode === 'multiple'
-            ? 1
-            : (theme) =>
+      sx={[
+        rootProps.rowGroupingColumnMode === 'multiple'
+          ? {
+              ml: 1,
+            }
+          : {
+              ml: (theme) =>
```
Comment (cherniavskii) [inline packages/x-data-grid-premium/src/components/GridGroupingColumnLeafCell.tsx:18]: Should use the `sx` callback signature instead?

File: packages/x-data-grid/src/components/panel/filterPanel/GridFilterForm.tsx [hunk 4]
```diff
@@ -452,12 +452,24 @@ const GridFilterForm = React.forwardRef<HTMLDivElement, GridFilterFormProps>(
...
+                },
+            showMultiFilterOperators
+              ? {
+                  visibility: 'visible',
+                }
+              : {
+                  visibility: 'hidden',
+                },
+            baseFormControlProps.sx || {},
+            logicOperatorInputProps.sx || {},
```
Comment (cherniavskii) [inline packages/x-data-grid/src/components/panel/filterPanel/GridFilterForm.tsx:471]:
```suggestion
            baseFormControlProps.sx,
            logicOperatorInputProps.sx,
```

File: packages/x-data-grid-premium/src/components/GridGroupingColumnLeafCell.tsx (hunk 3 above)
Comment (LukasTy) [reply packages/x-data-grid-premium/src/components/GridGroupingColumnLeafCell.tsx:18]: @siriwatknp confirmed that this result from codemod is correct. 🤔

Comment (cherniavskii) [reply packages/x-data-grid-premium/src/components/GridGroupingColumnLeafCell.tsx:18]: Isn't it the opposite of https://github.com/mui/mui-x/pull/14143#discussion_r1718271042?
Or am I missing something?

Comment (LukasTy) [reply packages/x-data-grid-premium/src/components/GridGroupingColumnLeafCell.tsx:18]: This looked strange to me as well. 🙈 
I'm asking @siriwatknp for a review of this PR to check these changes. 🤔

**Comment by oliviertassinari** [comment]
Will we need an helper like for v5? https://github.com/mui/mui-x/pull/2108/files#diff-b4487b0ccf1b0178520560e5e9b18efbd0d4030106ae794bda6638be8cbdd526L36

File: docs/src/modules/components/LocalizationTable.js [hunk 5]
```diff
@@ -29,26 +26,29 @@ const Root = styled('div')(({ theme }) => ({
       border: `1px solid`,
       borderColor: (theme.vars || theme).palette.error.light,
     },
+    ...theme.applyStyles('dark', {
```
Comment (oliviertassinari) [inline docs/src/modules/components/LocalizationTable.js:29]: I was curious about whether we could use `applyStyles()` or not. It was added with https://github.com/mui/material-ui/pull/40667, released with @mui/material@5.15.7, we have this dependency:

https://github.com/mui/mui-x/blob/c0b599300f01c54a96a4443a6296b10c6555036d/packages/x-data-grid/package.json#L61

since https://github.com/mui/mui-x/pull/12516, v7.0.0, so it looks likes yes. Cool

**Comment by LukasTy** [comment]
> Will we need an `isMuiV5()` helper like for v5? https://github.com/mui/mui-x/pull/2108/files#diff-b4487b0ccf1b0178520560e5e9b18efbd0d4030106ae794bda6638be8cbdd526L36

From the initial investigation—it doesn't look like we'll need it @oliviertassinari, because all the changes are supported by the `@mui/material` version we have listed in `peerDependencies`. 🤔

**Review by alexfauquette — APPROVED** [review]
CHnages on docs and charts look good to me

File: docs/src/modules/components/DemoPropsForm.tsx (hunk 2 above)
Comment (DiegoAndai) [reply docs/src/modules/components/DemoPropsForm.tsx:132]: AFAIK it's not included in the migration guide as it's covered by the codemod.

@siriwatknp, what do you think about adding an example for all cases covered by the codemod? Would there be too many?

File: packages/x-data-grid-premium/src/components/GridGroupingColumnLeafCell.tsx (hunk 3 above)
Comment (siriwatknp) [reply packages/x-data-grid-premium/src/components/GridGroupingColumnLeafCell.tsx:18]: Sorry, I missed one case. It should be:

```js
sx={[
  rootProps.rowGroupingColumnMode === 'multiple'
          ? {
              ml: 1,
            }
          : (theme) => ({
              ml: `calc(var(--DataGrid-cellOffsetMultiplier) * ${theme.spacing(rowNode.depth)})`
            })
]}
```

Comment (siriwatknp) [reply packages/x-data-grid-premium/src/components/GridGroupingColumnLeafCell.tsx:18]: > This looked strange to me as well

It looks strange to me at first but if you view it from static perspective, it starts to make sense.

Each style (an object of CSS) must be static, so callback as values are not allowed. However, each style can be a callback in order to access the theme (theme is static because it's configured at build time). So here is what style can be:

1. plain object: `{ display: 'flex', ml: 2, …etc }`
2. a callback that return an object: `theme => ({ display: 'flex', ml: theme.spacing(2), …etc })`

Each style generates a class name that will be used as a reference to the style.

Now the dynamic part. From static point of view, you cannot switch the value of a CSS property based on condition but what you can do is switching classes (styles) based on condition. That's why the result looks like:

`sx={condition ? { … } : { … }}` => transformed to `sx={condition ? 'classname1' : 'classname2'}`

The same concept applies for multiple styles using array:

```js
sx={[
  condition1 ? { … } : (theme => {…}),
  condition2 ? { … } : { … }
]}
```

Does this help?

Comment (siriwatknp) [inline packages/x-data-grid-premium/src/components/GridGroupingColumnLeafCell.tsx:18]:
```suggestion
              theme => ({ ml:
```

File: docs/data/date-pickers/custom-opening-button/AddWarningIconWhenInvalidRange.tsx (hunk 1 above)
Comment (siriwatknp) [reply docs/data/date-pickers/custom-opening-button/AddWarningIconWhenInvalidRange.tsx:26]: This is required. Please have a look at my [explanation](https://github.com/mui/mui-x/pull/14143/files/e21fdcf9baa322e13ba67c5c8c868c5b9287bf2c#r1723650800)

File: packages/x-data-grid-premium/src/components/GridGroupingColumnLeafCell.tsx (hunk 3 above)
Comment (DiegoAndai) [reply packages/x-data-grid-premium/src/components/GridGroupingColumnLeafCell.tsx:18]: @siriwatknp what about this:

```js
      sx={
        (theme) => ({
           ml: rootProps.rowGroupingColumnMode === 'multiple'
             ? 1
             : `calc(var(--DataGrid-cellOffsetMultiplier) * ${theme.spacing(rowNode.depth)})`
         })
       }
```

- Does this work?
- If it works, is there a preferred approach?

File: docs/src/modules/components/DemoPropsForm.tsx (hunk 2 above)
Comment (siriwatknp) [reply docs/src/modules/components/DemoPropsForm.tsx:132]: For small cases like this, I can update the migration page but it should not block the v6.

Comment (DiegoAndai) [reply docs/src/modules/components/DemoPropsForm.tsx:132]: Let's create an issue to add the additional examples, but I agree not to block the stable release on this one

**Review by siriwatknp — CHANGES_REQUESTED** [review]
Please change https://github.com/mui/mui-x/pull/14143/files#diff-d2872fe72d025095f6347b61c0ac177d659c43b2b669a58e9c935760fa1b602e to:

```js
<Box
   sx={[
        rootProps.rowGroupingColumnMode === 'multiple'
          ? {
              ml: 0,
            }
          : theme => ({
              ml: `calc(var(--DataGrid-cellOffsetMultiplier) * var(--_depth) * ${theme.spacing(1)})`,
            }),
      ]}
  style={{
    '--depth': rowNode.depth
  } as any}
>
```

File: docs/data/date-pickers/custom-opening-button/AddWarningIconWhenInvalidRange.tsx (hunk 1 above)
Comment (DiegoAndai) [reply docs/data/date-pickers/custom-opening-button/AddWarningIconWhenInvalidRange.tsx:26]: By the way, in cases such as this one in which there's one condition, you can remove the array

```js
        sx={
          hasError
            ? {
                opacity: 1,
              }
            : {
                opacity: 0,
              },
        }
```

File: packages/x-data-grid-premium/src/components/GridGroupingColumnLeafCell.tsx (hunk 3 above)
Comment (cherniavskii) [reply packages/x-data-grid-premium/src/components/GridGroupingColumnLeafCell.tsx:18]: Alright, I was confused by this: `ml: (theme) => {...}` 😅

The theme is only available when using the callback signature on the **style object level** 👍🏻

File: docs/src/modules/components/LocalizationTable.js (hunk 5 above)
Comment (cherniavskii) [reply docs/src/modules/components/LocalizationTable.js:29]: Yes, I was pretty sure we can't use it 😅
But it turns out the minimum supported version guarantees that `applyStyles` will always be available 👍🏻

File: packages/x-charts/src/ChartsAxisHighlight/ChartsAxisHighlight.tsx [hunk 6]
```diff
@@ -37,16 +37,34 @@ export const ChartsAxisHighlightPath = styled('path', {
...
-  }),
-  ...(ownerState.axisHighlight === 'line' && {
-    strokeDasharray: '5 2',
-    stroke: theme.palette.mode === 'light' ? '#000000' : '#ffffff',
-  }),
+  variants: [
+    {
+      props: {
+        axisHighlight: 'band',
+      },
```
Comment (cherniavskii) [inline packages/x-charts/src/ChartsAxisHighlight/ChartsAxisHighlight.tsx:46]: This doesn't seem right, since `axisHighlight` is part of the `ownerState` prop: https://github.com/mui/mui-x/blob/8a75b228b616b1dceb24f85721340db41fc48e8a/packages/x-charts/src/ChartsAxisHighlight/ChartsAxisHighlight.tsx#L116

@siriwatknp How should we approach this?
Would this work?

```
variants: [
  {
    props: {
      ownerState: {
        axisHighlight: 'band',
      }
    },
```

Or should we do this instead?

```diff
<ChartsAxisHighlightPath
  ownerState={{ axisHighlight: 'band' }}
+ axisHighlight="band"
/>
```

Comment (cherniavskii) [reply packages/x-charts/src/ChartsAxisHighlight/ChartsAxisHighlight.tsx:46]: Or this?

```
variants: [
  {
    props: ({ ownerState }) => ownerState.axisHighlight === 'band',
```

File: packages/x-charts/src/Gauge/GaugeContainer.tsx [hunk 7]
```diff
@@ -25,9 +25,9 @@ export interface GaugeContainerProps
 const ResizableContainer = styled('div', {
   name: 'MuiGauge',
   slot: 'Container',
-})<{ ownerState: Pick<GaugeContainerProps, 'width' | 'height'> }>(({ ownerState, theme }) => ({
-  width: ownerState.width ?? '100%',
-  height: ownerState.height ?? '100%',
+})<{ ownerState: Pick<GaugeContainerProps, 'width' | 'height'> }>(({ theme }) => ({
+  width: 'var(--width, 100%)',
+  height: 'var(--height, 100%)',
```
Comment (cherniavskii) [inline packages/x-charts/src/Gauge/GaugeContainer.tsx:30]: This seems fragile though:

```jsx
<div style={{ '--width': 500 }}>
  <div
    style={{
      // '--width': 200,
      
      // If forget to set the variable above, the width below won't fall back to `100%`
      // since CSS variables are inherited and the value of 500 will be used
      
      width: 'var(--width, 100%)',
    }}>
  </div>
</div>
```

Does this imply that CSS Variable names should be unique?
cc @siriwatknp @DiegoAndai

File: docs/data/charts/styling/SxStyling.js
Comment (cherniavskii) [inline docs/data/charts/styling/SxStyling.js:1]: There are some components that I don't know how to migrate, e.g. https://github.com/mui/mui-x/blob/8a75b228b616b1dceb24f85721340db41fc48e8a/packages/x-charts/src/BarChart/BarElement.tsx#L48-L59

Any ideas?
@LukasTy @siriwatknp @DiegoAndai

File: packages/x-charts/src/ChartsAxisHighlight/ChartsAxisHighlight.tsx (hunk 6 above)
Comment (DiegoAndai) [reply packages/x-charts/src/ChartsAxisHighlight/ChartsAxisHighlight.tsx]: @cherniavskii, this should work:

```js
variants: [
  {
    props: { axisHighlight: 'band' }
```

The owner state values are checked in the variant's props check ([source code](https://github.com/mui/material-ui/blob/next/packages/mui-system/src/createStyled/createStyled.js#L69)). The owner state values are also available in the prop callback ([source code](https://github.com/mui/material-ui/blob/next/packages/mui-system/src/createStyled/createStyled.js#L63-L64)). This is also what we're doing in Material UI.

Let me know if it doesn't work 😊.

File: packages/x-charts/src/Gauge/GaugeContainer.tsx (hunk 7 above)
Comment (DiegoAndai) [reply packages/x-charts/src/Gauge/GaugeContainer.tsx:30]: We haven't settled on a pattern in Material UI yet (will probably do for v7), but I would say yes, a unique name would be useful to avoid future bugs. In most instances, we're prefixing the component name to the variable, for example: https://github.com/mui/material-ui/blob/next/packages/mui-material/src/AppBar/AppBar.js#L100

File: docs/data/charts/styling/SxStyling.js
Comment (DiegoAndai) [reply docs/data/charts/styling/SxStyling.js:1]: I think this should work:

```js
export const BarElementPath = styled(animated.rect, {
  name: 'MuiBarElement',
  slot: 'Root',
  overridesResolver: (_, styles) => styles.root,
})<{ ownerState: BarElementOwnerState }>(() => ({
  stroke: 'none',
  transition: 'opacity 0.2s ease-in, fill 0.2s ease-in',
  opacity: 1, // might not be required, but I'm not familiar with the animation
  variants: [
    {
      props: { isHighlighted: true },
      style: ({ ownerState }) => ({
        fill: d3Color(ownerState.color)!.brighter(0.5).formatHex()
      })
    },
    {
      props: { isHighlighted: false },
      style: ({ ownerState }) => ({
        fill: ownerState.color
      })
    },
    {
      props: { isFaded: true },
      style: { opacity: 0.3 },
    },
  ],
}));
```

The `props` and `ownerState` are also available in the `style` callback, here's a demo: https://codesandbox.io/p/sandbox/mui-starter-ckpm5s

@siriwatknp can confirm if this is the correct usage.

File: packages/x-charts/src/Gauge/GaugeContainer.tsx [hunk 8]
```diff
@@ -77,6 +77,12 @@ const GaugeContainer = React.forwardRef(function GaugeContainer(props: GaugeCont
       aria-valuenow={value === null ? undefined : value}
       aria-valuemin={valueMin}
       aria-valuemax={valueMax}
+      style={
+        {
+          '--width': inWidth,
+          '--height': inHeight,
```
Comment (siriwatknp) [inline packages/x-charts/src/Gauge/GaugeContainer.tsx:83]:
```suggestion
          '--width': typeof inWidth === 'number' ? `${inWidth}px` : inWidth,
          '--height': typeof inHeight === 'number' ? `${inHeight}px` : inHeight,
```

as `width: 200` is invalid in CSS.

File: packages/x-charts/src/LineChart/LineHighlightElement.tsx [hunk 9]
```diff
@@ -43,11 +43,11 @@ const HighlightElement = styled('circle', {
...
   slot: 'Root',
   overridesResolver: (_, styles) => styles.root,
-})<{ ownerState: LineHighlightElementOwnerState }>(({ ownerState }) => ({
-  transform: `translate(${ownerState.x}px, ${ownerState.y}px)`,
-  transformOrigin: `${ownerState.x}px ${ownerState.y}px`,
-  fill: ownerState.color,
-}));
+})<{ ownerState: LineHighlightElementOwnerState }>({
+  transform: `translate(var(--x)px, var(--y)px)`,
+  transformOrigin: `var(--x)px var(--y)px`,
```
Comment (siriwatknp) [inline packages/x-charts/src/LineChart/LineHighlightElement.tsx:48]:
```suggestion
  transform: `translate(var(--x), var(--y))`,
  transformOrigin: `var(--x) var(--y)`,
```

concatenate won't work in CSS, please move to inline style.

File: packages/x-charts/src/LineChart/LineHighlightElement.tsx [hunk 10]
```diff
@@ -82,6 +82,14 @@ function LineHighlightElement(props: LineHighlightElementProps) {
       cx={0}
       cy={0}
       r={other.r === undefined ? 5 : other.r}
+      style={
+        {
+          ...other.style,
+          '--x': x,
+          '--y': y,
```
Comment (siriwatknp) [inline packages/x-charts/src/LineChart/LineHighlightElement.tsx:89]:
```suggestion
          '--x': `${x}px`,
          '--y': `${y}px`,
```

File: docs/src/modules/components/DemoPropsForm.tsx (hunk 2 above)
Comment (siriwatknp) [reply docs/src/modules/components/DemoPropsForm.tsx:132]: Sure, I can update the codemod to link to the migration as well if that helps.

File: docs/data/charts/styling/SxStyling.js
Comment (siriwatknp) [reply docs/data/charts/styling/SxStyling.js:1]: This won't work with Pigment, the `ownerState` usage is still dynamic. It should be:

```js
export const BarElementPath = styled(animated.rect, {
  name: 'MuiBarElement',
  slot: 'Root',
  overridesResolver: (_, styles) => styles.root,
})<{ ownerState: BarElementOwnerState }>(() => ({
  stroke: 'none',
  transition: 'opacity 0.2s ease-in, fill 0.2s ease-in',
  opacity: 1, // might not be required, but I'm not familiar with the animation
  fill: 'var(--_fill)', // the `_` is a convention in CSS to indicates that the variable is private.
  variants: [
    {
      props: { isFaded: true },
      style: { opacity: 0.3 },
    },
  ],
}));

<BarElementPath style={{
  '--_fill': isHighlighted ? d3Color(ownerState.color)!.brighter(0.5).formatHex() : color,
}}>
```

Comment (LukasTy) [reply docs/data/charts/styling/SxStyling.js:1]: @cherniavskii Supporting PigmentCSS is not in the scope of this PR.
I've only applied changes that were made by codemod, but for full support, we potentially need way more work and attention for it.
Personally, I would prefer omitting any manual changes for Pigment support to reduce the scope of the PR, need for review and the potential of errors. 🙈 
WDYT?

File: packages/x-charts/src/ChartsAxisHighlight/ChartsAxisHighlight.tsx (hunk 6 above)
Comment (LukasTy) [reply packages/x-charts/src/ChartsAxisHighlight/ChartsAxisHighlight.tsx]: AFAIK, I did thorough testing of the changes and my solution seemed to work.
It is done in multiple places on Pickers codebase.

**Review by siriwatknp — APPROVED** [review]

File: docs/data/charts/styling/SxStyling.js
Comment (LukasTy) [reply docs/data/charts/styling/SxStyling.js:1]: @cherniavskii I've removed the additional changes you introduced to avoid potential problems.
They would not be enough for PigmentCSS support anyways. 🙈 
I think that support for it should be handled separately.

Comment (cherniavskii) [reply docs/data/charts/styling/SxStyling.js:1]: Yes, the example I gave here was out of curiosity 👍🏻
